### DISCUSSION TITLE
Algorithmic reprocessing for RNG

### DISCUSSION THREAD

**Discussion by marcotrijulien** [strongswan/strongswan#2319] [Q&A] [answered]
Hello, I would like some information regarding how strongSwan RNG works. 

I am currently using this :
```
strongSwan 5.9.11 swanctl
loaded plugins: pkcs11 aes des rc2 sha2 sha1 md5 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs12 pgp dnskey sshkey pem openssl pkcs8 fips-prf curve25519 xcbc cmac hmac kdf gcm drbg
```

For cybersecurity reasons, I have to follow some rules concerning random number generators used for cryptographic purposes, for example there must be an algorithmic reprocessing, not just use the output of a source of randomness.

So, I would like to know if strongSwan just reads random numbers from _/dev/random_ or _/dev/urandom_ without any algorithmic reprocessing (like AES-CTR and/or HMAC-SHA2), or is it using the "drbg" plugin, which is loaded. Or how to use the drbg instead of just reading _/dev/random_.

I haven't changed strongswan.conf "charon.plugins.random" and here is what I get with `swanctl --list-algs` :
```
drbg:
  DRBG_CTR_AES128[drbg]
  DRBG_CTR_AES192[drbg]
  DRBG_CTR_AES256[drbg]
  DRBG_HMAC_SHA1[drbg]
  DRBG_HMAC_SHA256[drbg]
  DRBG_HMAC_SHA384[drbg]
  DRBG_HMAC_SHA512[drbg]
rng:
  RNG_WEAK[openssl]
  RNG_STRONG[random]
  RNG_TRUE[random]

```

**Comment by tobiasbrunner** [comment, marked as answer]
> So, I would like to know if strongSwan just reads random numbers from /dev/random or /dev/urandom without any algorithmic reprocessing (like AES-CTR and/or HMAC-SHA2),

It depends on the plugin configuration, in particular in relation to the `openssl` plugin, which is of particular interest due to its FIPS module. With other crypto plugins the random source can differ, e.g. `wolfssl` is similar to `openssl`, but the `botan` plugin can be configured to use strongSwan's `rng_t` provided by other plugins, or use its internal RNG only.

Assuming the `openssl` plugin is the **only crypto plugin loaded**:

 * If the `random` plugin is **not loaded**, then only the OpenSSL DRBG will be used.
 * If the `random` plugin is **loaded after** the `openssl` plugin, then the OpenSSL DRBG will be used for all random values as well.
 * If the `random` plugin is **loaded before** the `openssl` plugin, then `/dev/urandom` will be used via `RNG_STRONG` (e.g. for the secret for COOKIE notifies), while the `openssl` plugin will be used for nonces and other small random values via `RNG_WEAK` (which the `random` plugin does not explicitly provide because there is a fallback to `RNG_STRONG` if no implementation for `RNG_WEAK` is registered). However, for DH keys, the source will always be the OpenSSL DRBG as we use OpenSSL's API to generate these keys. 

In the latter two cases, the `random` plugin will still be registered for the `RNG_TRUE` class as the `openssl` plugin does not provide this class of RNGs.  It's used for long-term secrets like RSA keys when generated by e.g. the `gmp` plugin via `pki --gen`, and the `random` plugin would read from `/dev/random` for it.

What's important regarding this is that (1) the IKE daemon does not use the `RNG_TRUE` class at runtime, so it's fine if no implementation is registered, and (2), similar to the DH keys, if the `openssl` plugin is used to generate long-term secrets with `pki --gen`, this still uses the OpenSSL DRBG even if the `random` plugin is loaded.

So if you want to make sure that only OpenSSL's DRBG is used (e.g. in FIPS scenarios), you should make sure your plugin list reflects that and disable the plugins you don't want (such as the `random` plugin or redundant crypto plugins).

> or is it using the "drbg" plugin, which is loaded.

No, the only users of `DBRG_HMAC_*` are the `gmp` plugin (to reduce the entropy requirements a bit when generating RSA key pairs) and the legacy `ntru` plugin.  And `DRBG_CTR_*` is only used by strongSwan 6.0 (for PQC test vectors and the `frodo` plugin).

**Reply by marcotrijulien** [reply]
I have removed the "random" plugin, I now have this :

```
strongSwan 5.9.11 swanctl
loaded plugins: pkcs11 aes des rc2 sha2 sha1 md5 nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs12 pgp dnskey sshkey pem openssl pkcs8 fips-prf curve25519 xcbc cmac hmac kdf gcm dr bg
```

But when checking with `swanctl --list-algs`, I get the same result: 

```
rng:
  RNG_WEAK[openssl]
  RNG_STRONG[random]
  RNG_TRUE[random]
```

Is it really using openssl DRBG ?

**Reply by tobiasbrunner** [reply]
Note that the plugins in swanctl are not necessarily the same plugins as those in the IKE daemon (they are configured separately). The output of `swanctl --help` will list the plugins loaded by the swanctl tool, while the `swanctl --stats` or `swanctl --list-algs` commands are communicating with the IKE daemon and list plugins and features loaded by the latter.

**Reply by marcotrijulien** [reply]
When doing `swanctl --stats`:
```
uptime: 2 hours, since Jul 02 11:54:55 2024
worker threads: 16 total, 11 idle, working: 4/0/1/0
job queues: 0/0/0/0
jobs scheduled: 3
IKE_SAs: 1 total, 0 half-open
mallinfo: sbrk 860160, mmap 0, used 708872, free 151288
loaded plugins: charon pkcs11 random aes des rc2 sha2 sha1 md5 nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs12 pgp dnskey sshkey pem openssl pkcs8 fips-prf curve25519 xcbc cmac hmac kdf gcm drbg attr kernel-netlink resolve socket-default stroke vici updown xauth-generic led counters
```

So, from what I understand, if I want to use OpenSSL DRBG for my VPNs I have to remove "random" from there, how ? I've already done the `--disable-random` option when compiling strongswan.

Is checking with `swanctl --list-algs` that I get something like this 👇  a good way to know if I use the OpensSSL DRBG ?
```
rng:
  RNG_WEAK[openssl]
  RNG_STRONG[openssl]
```

Is there any **strongswan.conf** modifications I can do to force it that way ?

**Reply by tobiasbrunner** [reply]
> I've already done the --disable-random option when compiling strongswan.

If you enabled it before, then you need to make sure the plugin file is not still installed and no snippet is in `/etc/strongswan.d/charon` that causes the installed plugin to get loaded.

> Is checking with `swanctl --list-algs` that I get something like this 👇 a good way to know if I use the OpensSSL DRBG ?

Sure.

> Is there any **strongswan.conf** modifications I can do to force it that way ?

You can disable plugins explicitly by disabling their `load` option. See [the docs](https://docs.strongswan.org/docs/5.9/plugins/pluginLoad.html).

**Reply by marcotrijulien** [reply]
Nice, thank you I finally get the result I want with `swanctl --list-algs`, there was the file `/etc/strongswan.d/charon/random.conf` which loaded the "random" plugin.
